File: python/src/algopay/payment/batch.py
"""Batch payment processor."""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING

from algopay.core.types import (
    BatchPaymentResult,
    PaymentMethod,
    PaymentRequest,
    PaymentResult,
    PaymentStatus,
)

if TYPE_CHECKING:
    from algopay.payment.router import PaymentRouter


class BatchProcessor:
    def __init__(self, router: PaymentRouter) -> None:
        self._router = router

    async def process(self, requests: list[PaymentRequest], concurrency: int = 5) -> BatchPaymentResult:
        sem = asyncio.Semaphore(concurrency)

        async def _bounded_pay(req: PaymentRequest) -> PaymentResult:
            async with sem:
                return await self._router.pay(
                    wallet_id=req.wallet_id,
                    recipient=req.recipient,
                    amount=req.amount,
                    purpose=req.purpose,
                    idempotency_key=req.idempotency_key,
                    destination_chain=req.destination_chain,
                    **req.metadata,
                )

        results = await asyncio.gather(*[_bounded_pay(req) for req in requests], return_exceptions=True)

        final: list[PaymentResult] = []
        for i, res in enumerate(results):
            if isinstance(res, PaymentResult):
                final.append(res)
            else:
                req = requests[i]
                final.append(
                    PaymentResult(
                        success=False,
                        transaction_id=None,
                        blockchain_tx=None,
                        amount=req.amount,
                        recipient=req.recipient,
                        method=PaymentMethod.TRANSFER,
                        status=PaymentStatus.FAILED,
                        error=str(res),
                    )
                )

        success_count = sum(1 for r in final if r.success)
        tx_ids = [r.transaction_id for r in final if r.transaction_id]
        return BatchPaymentResult(
            total_count=len(final),
            success_count=success_count,
            failed_count=len(final) - success_count,
            results=final,
            transaction_ids=[t for t in tx_ids if t],
        )

File: typescript/src/payment/batch.ts
import { PaymentMethod, PaymentStatus, type PaymentRequest, type PaymentResult } from "../types.js";
import type { PaymentRouter } from "./router.js";

export interface BatchPaymentResult {
  totalCount: number;
  successCount: number;
  failedCount: number;
  results: PaymentResult[];
  transactionIds: string[];
}

async function withConcurrency<T>(
  tasks: (() => Promise<T>)[],
  concurrency: number,
): Promise<PromiseSettledResult<T>[]> {
  const results: PromiseSettledResult<T>[] = new Array(tasks.length);
  let idx = 0;

  async function worker(): Promise<void> {
    while (idx < tasks.length) {
      const i = idx++;
      try {
        results[i] = { status: "fulfilled", value: await tasks[i]() };
      } catch (e) {
        results[i] = { status: "rejected", reason: e };
      }
    }
  }

  const workers = Array.from({ length: Math.min(concurrency, tasks.length) }, () => worker());
  await Promise.all(workers);
  return results;
}

export class BatchProcessor {
  private readonly _router: PaymentRouter;

  constructor(router: PaymentRouter) {
    this._router = router;
  }

  async process(requests: PaymentRequest[], concurrency = 5): Promise<BatchPaymentResult> {
    const tasks = requests.map(
      (req) => () =>
        this._router.pay({
          walletId: req.walletId,
          recipient: req.recipient,
          amount: req.amount,
          purpose: req.purpose,
          feeLevel: undefined,
        }),
    );

    const settled = await withConcurrency(tasks, concurrency);

    const results: PaymentResult[] = settled.map((s, i) => {
      if (s.status === "fulfilled") return s.value;
      const req = requests[i];
      return {
        success: false,
        transactionId: null,
        blockchainTx: null,
        amount: req.amount,
        recipient: req.recipient,
        method: PaymentMethod.TRANSFER,
        status: PaymentStatus.FAILED,
        error: String(s.reason),
      };
    });

    const successCount = results.filter((r) => r.success).length;
    const transactionIds = results
      .map((r) => r.transactionId)
      .filter((id): id is string => id != null);

    return {
      totalCount: results.length,
      successCount,
      failedCount: results.length - successCount,
      results,
      transactionIds,
    };
  }
}
